Go tutorial 🐿️ (#8904)
Unleash/unleash: website/docs/feature-flag-tutorials/golang/implementing-feature-flags-golang.md

---
title: How to Implement Feature Flags in Go
description: "How to use Unleash feature flags with Go."
slug: /feature-flag-tutorials/go
---

> import VideoContent from '@site/src/components/VideoContent.jsx';

Reviewer: Is this supposed to be here?

```
git clone https://github.com/unleash/unleash.git
cd unleash
docker compose up -d
```

You will now have Unleash installed onto your machine and running in the background. You can access this instance in your web browser at [http://localhost:4242](http://localhost:4242).

Log in to the platform using these credentials:

```
Username: admin
Password: unleash4all
```

Click **New feature flag** to create a new feature flag.

![Create a new feature flag](/img/go-new-feature-flag.png)

Call it `graphql-api` and enable it in the `development` environment.

![A feature flag called `graphql-api` is now enabled in development.](/img/go-enable-development.png)

Everything's now setup on the Unleash side. Let's go to the code now.

## 3. Grab country information from the REST Countries API

We'll use the standard `net/http` package to make our HTTP requests and the Unleash SDK to connect to your local Unleash instance and retrieve your feature flag.

Open a new tab in your terminal, and create a new folder (NOT in the unleash folder).

```sh
mkdir unleash-go
cd unleash-go
go mod init unleash-demo
```

Install dependencies:

```sh
go get github.com/Unleash/unleash-client-go/v4
```

> Then, let's make sure our setup is working. Let's make a call to the REST API to retrieve information about a country using its country code. Create a file named `main.go`:

Reviewer 2: ```suggestion
Next, let's make sure our setup is working. Make a call to the REST API to retrieve information about a country using its country code. Create a file named `main.go`:
```

In this tutorial, we'll get information about a country from the [Countries API](https://restcountries.com/), using just Go. We'll use feature flags to decide whether to call the REST or the GraphQL version of the API.

- [Prerequisites](#prerequisites)
- [1. Best practices for back-end apps with Unleash](#1-best-practices-for-back-end-apps-with-unleash)
- [2. Install a local feature flag provider](#2-install-a-local-feature-flag-provider)
- [3. Grab country information from the REST Countries API](#3-grab-country-information-from-the-rest-countries-api)
- [4. Add the GraphQL endpoint](#4-add-the-graphql-endpoint)
- [5. Add Unleash to your Go app](#5-add-unleash-to-your-go-app)
- [6. Verify the toggle experience](#6-verify-the-toggle-experience)
- [Conclusion](#conclusion)

## Prerequisites

For this tutorial, you'll need the following:

-   Go v1.16+
-   Git
-   Docker and Docker Compose

> -   (Optional) `go mod`, to manage your dependencies

Reviewer 3: ```suggestion
-   Go Modules, to manage your dependencies
```

![architecture diagram for our implementation](./diagram.png)

> The Unleash Server is a **Feature Flag Control Service**, which manages your feature flags and lets you retrieve flag data. Unleash has a UI for creating and managing projects and feature flags. There are also [API commands available](https://docs.getunleash.io/reference/api/unleash) to perform the same actions straight from your CLI or server-side app.

Reviewer 3: **Feature Flag Control Service** I think refers to a different diagram

## 2. Install a local feature flag provider

In this section, we'll install Unleash, run the instance locally, log in, and create a feature flag. If you prefer, you can use other tools instead of Unleash, but you'll need to update the code accordingly. The basic steps will probably be the same.

Use Git to clone the Unleash repository and Docker to build and run it. Open a terminal window and run the following commands:

```
git clone https://github.com/unleash/unleash.git
cd unleash
docker compose up -d
```

You will now have Unleash installed onto your machine and running in the background. You can access this instance in your web browser at [http://localhost:4242](http://localhost:4242).

Log in to the platform using these credentials:

```
Username: admin
Password: unleash4all
```

Click the 'New feature flag' button to create a new feature flag.

![Create a new feature flag](../ruby/new-ff.png)

Call it `graphql-api` and enable it in the `development` environment.

![A feature flag called `graphql-api` is now visible.](../ruby/enable-ff.png)

> Everything's now setup on the Unleash side. Let's go to the code now.

Reviewer 3: 😅 sorry
```suggestion
Everything's now set up on the Unleash side. Let's go to the code now.
```

## 3. Grab country information from the REST Countries API

We'll use the standard `net/http` package to make our HTTP requests and the Unleash SDK to connect to your local Unleash instance and retrieve your feature flag.

> Open a new tab in your terminal, and create a new folder (NOT in the unleash folder).

Reviewer 2: "outside **of** the unleash folder" right?"

```sh
mkdir unleash-go
cd unleash-go
go mod init unleash-demo
```

Install dependencies:

```sh
go get github.com/Unleash/unleash-client-go/v4
```

Then, let's make sure our setup is working. Let's make a call to the REST API to retrieve information about a country using its country code. Create a file named `main.go`:

```go
package main

import (
    "encoding/json"
    "fmt"
    "io"
    "log"
    "net/http"
)

type Country struct {
    Name    string `json:"name"`
    Capital string `json:"capital"`
}

func main() {
    resp, err := http.Get("https://restcountries.com/v2/alpha/no")
    if err != nil {
        log.Fatal(err)
    }
    defer resp.Body.Close()

    body, _ := io.ReadAll(resp.Body)
    var country Country
    json.Unmarshal(body, &country)

    fmt.Printf("Country: %s, Capital: %s\n", country.Name, country.Capital)
}
```

Run the code:

```sh
go run main.go
```

You should see `Country: Norway, Capital: Oslo` in your terminal.

## 4. Add the GraphQL endpoint

> The point of this tutorial is to mimic a real-world scenario where, based on a boolean feature flag, you would migrate from a REST API to a GraphQL one. So far, we've just used REST. Now, let's add the GraphQL version. The GraphQL endpoint is `https://countries.trevorblades.com/` (Thanks Trevor!).

Reviewer 3: Should we introduce the GraphQL API we're using at the same time as we introduce the REST API?

> Let's create a static feature flag, for now, just to test that we can call both versions successfully. Update your `main.go`:

Reviewer 3: ```suggestion
Let's create a static feature flag, for now, just to test that we can call both versions successfully. Update `main.go`:
```

            log.Fatal(err)
        }
        req.Header.Set("Content-Type", "application/json")

        client := &http.Client{}
        resp, err := client.Do(req)
        if err != nil {
            log.Fatal(err)
        }
        defer resp.Body.Close()

        body, _ := io.ReadAll(resp.Body)
        var response struct {
            Data struct {
                Country Country `json:"country"`
            } `json:"data"`
        }
        json.Unmarshal(body, &response)

        country = response.Data.Country
        fmt.Println("Hello GraphQL")
    } else {
        // Call the REST API
        resp, err := http.Get("https://restcountries.com/v2/alpha/no")
        if err != nil {
            log.Fatal(err)
        }
        defer resp.Body.Close()

        body, _ := io.ReadAll(resp.Body)
        var countries []Country
        json.Unmarshal(body, &countries)
        country = countries[0]
    }

    fmt.Printf("Country: %s, Capital: %s\n", country.Name, country.Capital)
}
```

Run the code again:

```sh
go run main.go
```

You should see `Hello GraphQL`, followed by `Country: Norway, Capital: Oslo` in your terminal.

## 5. Add Unleash to your Go app

> Now, let's connect our project to Unleash so that you can toggle that feature flag at runtime. If you wanted to, you could also do a gradual rollout, use it for a/b testing, etc.

Reviewer 2: I wonder if linking to the gradual rollout process or a/b testing would be good here.

You'll need 2 things:

-   The URL of your Unleash instance's API. It's `http://localhost:4242/api/` for your local version. You'll want to replace it with your remote instance.
-   The API token we created on our Unleash instance, feel free to create another one if you can't find it.

With these 2, you can initialize your Unleash client as follows:

```go
unleash.Initialize(
    unleash.WithUrl("http://localhost:4242/api/"),
    unleash.WithAppName("country_go"),
    unleash.WithCustomHeaders(http.Header{"Authorization": {"YOUR_API_KEY"}}),
)
```

You can check our [API token and client keys documentation](https://docs.getunleash.io/reference/api-tokens-and-client-keys) for more specifics.

Now, let's add our client to our project, grab the feature flag from Unleash, and update our conditional statement. Don't forget to also update the config with your API key:

> ```diff

Reviewer 3: How is this supposed to render? I don't think it works at the moment

Author: You're right! This was a problem with `diff` blocks across the board, fixed it 😄 https://github.com/Unleash/unleash/pull/8937

+		unleash.WithUrl("http://localhost:4242/api/"),
+		unleash.WithAppName("country_go"),
+		unleash.WithCustomHeaders(http.Header{"Authorization": {"YOUR_API_KEY"}}),
+	)
+
+	unleash.WaitForReady()

+	isGraphQL := unleash.IsEnabled("graphql-api")
-	// Define a static feature flag
-	isGraphQL := true

    // ... rest of the code ...
}
```

See additional use cases in our [Server-Side SDK with Go](https://docs.getunleash.io/reference/sdks/go) documentation.

## 6. Verify the toggle experience

Now that we've connected our project to Unleash and grabbed our feature flag, we can verify that if you disable that flag in your development environment, you stop seeing the `Hello GraphQL` message and only get the country information from the REST API.

> **Note:** An update to a feature flag may take 30 seconds to propagate.

![A feature flag called `graphql-api` is now disabled](../ruby/graphql-ff.png)

## Conclusion

All done! Now you know how to add feature flags with Unleash in Go. You've learned how to:

-   Toggle between a REST and a GraphQL endpoint based on a feature flag
-   Install Unleash and create/enable a feature flag
-   Grab the value of a feature flag with the Go SDK

> Thank you

Reviewer 2: We can link to the "Examples" doc here for next steps.
